File: Amazon/565.array-nesting.go
package leetcode

// ArrayNesting565 follows permutation links and marks visited indices by
// writing -1, which is outside the valid value range 0..n-1. Each index belongs
// to exactly one cycle/component.
//
// The input slice is mutated to keep O(1) auxiliary space.
// Time: O(n). Space: O(1).
func ArrayNesting565(nums []int) int {
	best := 0
	for i := range nums {
		if nums[i] == -1 {
			continue
		}
		count, j := 0, i
		for nums[j] != -1 {
			next := nums[j]
			nums[j] = -1
			j = next
			count++
		}
		best = maxInt(best, count)
	}
	return best
}

File: Amazon/565.array-nesting.java
/**
 * Algorithm:
 * Each index belongs to one cycle in the permutation-like array. Walk each
 * unvisited cycle, mark visited entries in-place as -1, and track the largest
 * cycle length.
 *
 * Complexity:
 * Time O(n), space O(1) because the input array is reused for visited marks.
 */
class Solution {
    public int arrayNesting(int[] nums) {
        int best = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == -1) {
                continue;
            }
            int count = 0;
            int j = i;
            while (nums[j] != -1) {
                int next = nums[j];
                nums[j] = -1;
                j = next;
                count++;
            }
            best = Math.max(best, count);
        }
        return best;
    }
}

